### PR TITLE
[SYCL][Graph] Add implicit queue recording mechanism

### PR DESCRIPTION
**Pull request by fabiomestre**
Updates the semantics of the record & replay API to allow queues to be
implicitly set to recording mode when submitting command-groups
that contains dependencies from other recording queues.

### CODE REVIEW

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -1499,14 +1499,39 @@ void handler::depends_on(const std::vector<event> &Events) {
 }
 
 void handler::depends_on(const detail::EventImplPtr &EventImpl) {
+
```
Comment (Bensuo) [inline sycl/source/handler.cpp:1502]: Unrelated whitespace change?

File: sycl/source/handler.cpp [hunk 2]
```diff
@@ -1499,14 +1499,39 @@ void handler::depends_on(const std::vector<event> &Events) {
...
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "Cannot submit to a queue with a dependency from a graph that is "
+          "associated with a different context.");
+      if (EventGraph->getDevice() != MQueue->get_device()) {
+        throw sycl::exception(
+            make_error_code(errc::invalid),
+            "Cannot submit to a queue with a dependency from a graph that is "
+            "associated with a different device.");
+      }
```
Comment (Bensuo) [inline sycl/source/handler.cpp:1522]: `begin_recording()` also checks if the queue is in fusion mode, do we not need to replicate that here?

Also could these checks be done in the call to `beginRecording()`?

File: sycl/source/detail/graph_impl.cpp [hunk 3]
```diff
@@ -1601,16 +1610,14 @@ void modifiable_command_graph::begin_recording(
                           "differs from the graph device.");
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp:1610]: Unrelated to this PR specifically, but it seems we have duplicate error checks here?

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  // Q2 is now in recording mode. Submitting a command group to Q3 with a
+  // dependency on an event from Q2 should change it to recording mode as well.
+  auto GraphEventD = Q3.submit([&](handler &CGH) {
+    CGH.depends_on(GraphEventB);
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrC[Id]++; });
+  });
+
+  assert(Q1.ext_oneapi_get_state() == exp_ext::queue_state::recording);
+  assert(Q2.ext_oneapi_get_state() == exp_ext::queue_state::recording);
+  assert(Q3.ext_oneapi_get_state() == exp_ext::queue_state::recording);
```
Comment (Bensuo) [inline sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp:68]: Slight nit but for this kind of state check I would probably explicitly check all 3 after each call to submit that the state is what it should be. Only checking all of them at the end leaves a bit of space for incorrect behaviour to sneak through.

This might also be better suited as a unit test, do we need to be running kernels on device for this test to function correctly?

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  Graph.begin_recording(Q1);
+
+  auto GraphEventA = Q1.submit([&](handler &CGH) {
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrA[Id]++; });
+  });
+
+  // Since there is a dependency on GraphEventA which is part of a graph,
+  // this will change Q2 to the recording state.
+  auto GraphEventB = Q2.submit([&](handler &CGH) {
+    CGH.depends_on(GraphEventA);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp:49]: Could we have a test that uses the queue `parallel_for` shortcuts too

File: sycl/source/detail/graph_impl.hpp [hunk 6]
```diff
@@ -848,7 +848,7 @@ class partition {
 };
 
 /// Implementation details of command_graph<modifiable>.
-class graph_impl {
+class graph_impl : public std::enable_shared_from_this<graph_impl> {
```
[the remark this answers is not in the document]
Comment (EwanC) [reply sycl/source/detail/graph_impl.hpp:851]: I can find two places where we pass in the shared_ptr of `this` as a  function parameter, e.g. https://github.com/intel/llvm/blob/sycl/sycl/source/detail/graph_impl.hpp#L956 & https://github.com/intel/llvm/blob/sycl/sycl/source/detail/graph_impl.hpp#L120

[the remark this answers is not in the document]
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:851]: That's fair, I'm fine with this then. We should track replacing those other instances for separate work though.

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp (hunk 4 above)
Comment (fabiomestre) [reply sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp:71]: I think running kernels can help make sure that the feature is working properly. Moving this to a unit test would only check if the queue state changed. But I think that even though the queue state has been updated correctly, the implementation can still be wrong.

I agree that some extra tests as unit tests might be useful though. In particular, I should add negative tests

File: sycl/source/detail/graph_impl.cpp (hunk 3 above)
Comment (fabiomestre) [reply sycl/source/detail/graph_impl.cpp]: Removed the duplicated checks :)

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp (hunk 5 above)
Comment (fabiomestre) [reply sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp:64]: Added a unit-test for the queue shortcuts

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  using T = int;
+
+  device Dev;
+  context Ctx{Dev};
+  queue Q1{Ctx, Dev};
+  queue Q2{Ctx, Dev};
+  queue Q3{Ctx, Dev};
+
+  exp_ext::queue_state Recording = exp_ext::queue_state::recording;
+  exp_ext::queue_state Executing = exp_ext::queue_state::executing;
```
Comment (EwanC) [inline sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp:24]: nitpick - Would stick a `const` on these since we're changing them.
```suggestion
  const exp_ext::queue_state Recording = exp_ext::queue_state::recording;
  const exp_ext::queue_state Executing = exp_ext::queue_state::executing;
```

**Review by EwanC — APPROVED** [review]
LGTM, nice testing too

File: sycl/source/handler.cpp (hunk 2 above)
Comment (Bensuo) [reply sycl/source/handler.cpp]: That's fair we can keep the errors separate for clarity.

File: sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp (hunk 4 above)
Comment (Bensuo) [reply sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp]: Having both is a good idea and unit tests look good to me!

**Review by Bensuo — APPROVED** [review]

**Review by againull — APPROVED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 8]
```diff
@@ -1453,6 +1453,38 @@ graph LR
...
+Submitting a command-group to a queue in the executable state can implicitly
+change its state to `queue_state::recording`. This will occur when the
+command-group depends on an event that has been returned by a queue in the
+recording state. The change of state happens before the command-group is
+submitted to the device (i.e. a new graph node will be created for that command-group).
+
+A queue whose state has been set to `queue_state::recording` using this
+mechanism, will behave as if it had been passed as an argument to
+`command_graph::begin_recording()`. In particular, its state will not
+change again until `command_graph::end_recording()` is called.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1467]: Making a note here that we should say something about the properties that we're passing by `begin_recording()`, if they are or aren't passed to the implicitly recorded queue. Tending towards that they are passed to queue being implicitly recording.

Comment (fabiomestre) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Added some extra wording for this behaviour.

Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: LGTM, thanks

**Comment by fabiomestre** [comment]
@intel/llvm-gatekeepers This should be ready to be merged.
